data/crac-creation/crac-creator-cim/src/main/java/com/powsybl/openrao/data/craccreation/creator/cim/craccreator/contingency/CimContingencyCreator.java

```diff
@@ -50,17 +48,31 @@ public void createAndAddContingencies() {
         this.cimContingencyCreationContexts = new HashSet<>();
 
         for (TimeSeries cimTimeSerie : cimTimeSeries) {
-            for (SeriesPeriod cimPeriodInTimeSerie : cimTimeSerie.getPeriod()) {
-                for (Point cimPointInPeriodInTimeSerie : cimPeriodInTimeSerie.getPoint()) {
-                    for (Series cimSerie : cimPointInPeriodInTimeSerie.getSeries().stream().filter(this::describesContingencyToImport).toList()) {
-                        for (ContingencySeries cimContingency : cimSerie.getContingencySeries()) {
-                            addContingency(cimContingency);
-                        }
+            addContingenciesIfTimestampInPeriod(cimTimeSerie);
+        }
+        this.cracCreationContext.setContingencyCreationContexts(cimContingencyCreationContexts);
+    }
+
+    private void addContingenciesIfTimestampInPeriod(TimeSeries cimTimeSerie) {
+        final java.time.Instant timestamp = cracCreationContext.getTimeStamp().toInstant();
+        final Comparator<Point> reversePointComparator = CimCracUtils.getReversePointComparator();
+        cimTimeSerie.getPeriod().forEach(
+            period -> {
+                List<Point> points = period.getPoint();
+                points.sort(reversePointComparator);
+                Optional<Integer> previousPosition = Optional.empty();
+                for (Point point : points) {
+                    final int currentPosition = point.getPosition();
+                    if (CimCracUtils.isTimestampInPeriod(timestamp, cimTimeSerie, period, currentPosition, previousPosition)) {
+                        point.getSeries().stream().filter(this::describesContingencyToImport).forEach(
```

Comment:
maybe add a method in CimCracUtils that returns a Set of Series, and use it in 3 "...Creator" classes, instead of duplicating the code. or do this filtering once before entering the classes

Reply:
the lines are not exactly the same, but the loops are the same. and since it's a pretty complicated code in the loops, I prefer to have it in one unique location